### PR TITLE
Throw Exceptions for Request Errors

### PR DESCRIPTION
**Pull request by bartboy011**
This PR introduces threw new exception classes:

- `TimeoutException` (cURL timeout)
- `ClientException` (`4**` responses)
- `ServerException` (`5**` responses)

Each exception contains the appropriate error message (ie, cURL's error or the response error) as well as the returned status code as the exception's code (except for TimeoutException).

All three exceptions extend the `RequestException` class, which extends the base `\Exception` class.

Additionally, this PR adds a cURL connection timeout of 10 seconds.

cc: @mthommes @cristiangrama

### CODE REVIEW

**Comment by bartboy011** [comment]
@ActiveCampaign/integration

File: includes/Connector.class.php
Comment (mthommes) [inline includes/Connector.class.php]: Could we make this comment a little more descriptive? I had to look up the exact difference between `DEFAULT_TIMEOUT` and `DEFAULT_CONNECTTIMEOUT`. For `DEFAULT_TIMEOUT` we could maybe say, "Default curl transfer timeout (before connection is established)" and for `DEFAULT_CONNECTTIMEOUT` we could say "Default curl connection timeout (after connection is established)."

That is... if I am understanding and phrasing that right.

File: includes/Connector.class.php
Comment (mthommes) [inline includes/Connector.class.php]: Can you add doc comments?

File: includes/Connector.class.php
Comment (mthommes) [inline includes/Connector.class.php]: Would we care `if curl_errno($request) && (string)curl_errno($request) !== '28'`? Because we don't capture that scenario.

**Review by mthommes — CHANGES_REQUESTED** [review]
Looks good! I just had a couple questions below. Also, this will require a major version change, right?

File: includes/Connector.class.php
Comment (bartboy011) [reply includes/Connector.class.php]: Good point, I'll make this clearer. I think you swapped your two descriptions around, `CONNECTTIMEOUT` is time spent waiting before a connection is made (if the server is 100% down, this will handle the timeout), `TIMEOUT` is time spent transferring data after the connection is made (if the server is overwhelmed and can't process the request, this will handle the timeout).

File: includes/Connector.class.php
Comment (bartboy011) [reply includes/Connector.class.php]: 👍

File: includes/Connector.class.php
Comment (bartboy011) [reply includes/Connector.class.php]: Yes, we probably would. I'll use the generic `RequestException` here

Comment (cristiangrama) [reply includes/Connector.class.php]: @bartboy011 
@mthommes brings in a good point. We should handle timeouts, 4xx and 5xx as well as other curl errors.

**Comment by bartboy011** [comment]
@mthommes Updated!

Edit: also @cristiangrama

**Review by mthommes — APPROVED** [review]
👍

**Comment by bartboy011** [comment]
@cristiangrama Looks like you don't get assigned after approval on this repo yet

**Comment by bartboy011** [comment]
Also, @mthommes to answer your question about the major version, yes, the next few PRs I'll be opening will all be part of v3

File: includes/Connector.class.php
Comment (cristiangrama) [inline includes/Connector.class.php]: CURLE_OPERATION_TIMEDOUT
Let's use the curl constant rather than the number. Also maybe a switch inside the first `if` would make it clearer. Also the comment needs to reflect the block of code.

Comment (bartboy011) [reply includes/Connector.class.php]: @cristiangrama Updated!

**Review by cristiangrama — APPROVED** [review]

**Pull request merged**
